CSHARP-2514: Support postBatchResumeToken in change streams.

DmitryLukyanov opened the pull request:
NOTE: This SPEC ticket is still in `open` state: https://github.com/mongodb/specifications/pull/533
which means that some logic can be changed.

DmitryLukyanov commented:
Evergreen: https://evergreen.mongodb.com/version/5caf94c79ccd4e48928a9a5b

DmitryLukyanov commented on src/MongoDB.Driver.Core/Core/Operations/AsyncCursor.cs:
```diff
@@ -120,6 +148,22 @@ public IEnumerable<TDocument> Current
+        {
+            get
+            {
+                var batch = _firstBatch ?? _currentBatch;
```
The spec line: 
`The initial aggregate response did not include a documentResumeToken (indicating no results).`
I guess having `isEmpty` flag is easier than additional document parsing, since all required varaibles to calculate batch size is already presented.

rstam replied:
As written this property has odd semantics. If called before MoveNext is ever called it returns whether the first batch is empty. If called after MoveNext is called it returns whether the *current* batch is empty. So initially it is looking forward and subsequently it is looking backwards.

Since we only need this to tell if the first batch was empty let's rename and simplify to:

    public bool WasFirstBatchEmpty => _wasFirstBatchEmpty;

And initialize _wasFirstBatchEmpty in the constructor:

     _wasFirstBatchEmpty = firstBatch.Count == 0;

DmitryLukyanov replied:
Done

DmitryLukyanov commented on src/MongoDB.Driver/IMongoClient.cs:
```diff
@@ -193,7 +193,7 @@ public interface IMongoClient
         /// </returns>
         Task<IAsyncCursor<BsonDocument>> ListDatabasesAsync(
-            IClientSessionHandle session,        
+            IClientSessionHandle session,
```
A lot of changes in `watch` methods are just about changing the return type

rstam replied:
Ack

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Core/Operations/ChangeStreamCursorTests.cs:
```diff
@@ -155,13 +211,154 @@ public void Dispose_can_be_called_more_than_once()
+
+        [SkippableTheory]
+        [ParameterAttributeData]
+        public void GetResumeToken_should_return_expected_results_when_batch_is_empty_or_fully_iterated(
```
Prose tests:
11.
 **For a ChangeStream under these conditions:**
            Running against a server >=4.0.7.
            The batch is empty or has been iterated to the last document.
 **Expected result:**
            getResumeToken must return the postBatchResumeToken from the current command response.

12.
 **For a ChangeStream under these conditions:**
            Running against a server <4.0.7.
            The batch is empty or has been iterated to the last document.
 **Expected result:**
            getResumeToken must return the _id of the last document returned if one exists.
            getResumeToken must return startAfter from the initial aggregate if the option was specified.
            getResumeToken must return resumeAfter from the initial aggregate if the option was specified.
            If neither the startAfter nor resumeAfter options were specified, the getResumeToken result must be empty.

DmitryLukyanov commented on tests/MongoDB.Driver.Core.Tests/Core/Operations/ChangeStreamCursorTests.cs:
```diff
@@ -155,13 +211,154 @@ public void Dispose_can_be_called_more_than_once()
+
+        [SkippableTheory]
+        [ParameterAttributeData]
+        public void GetResumeToken_should_return_expected_results_when_batch_is_not_empty_and_has_not_been_iterated(
```
Prose tests:
14. 
 **For a ChangeStream under these conditions:**
            The batch is not empty.
            The batch hasn’t been iterated at all.
            Only the initial aggregate command has been executed.
**Expected result:**
            getResumeToken must return startAfter from the initial aggregate if the option was specified.
            getResumeToken must return resumeAfter from the initial aggregate if the option was specified.
            If neither the startAfter nor resumeAfter options were specified, the getResumeToken result must be empty.

15.    
**For a ChangeStream under these conditions:**
            Running against a server >=4.0.7.
            The batch is not empty.
            The batch hasn’t been iterated at all.
            The stream has iterated beyond a previous batch and a getMore command has just been executed.

**Expected result:**
            getResumeToken must return the postBatchResumeToken from the previous command response.

16.
**For a ChangeStream under these conditions:**
            Running against a server <4.0.7.
            The batch is not empty.
            The batch hasn’t been iterated at all.
            The stream has iterated beyond a previous batch and a getMore command has just been executed.
**Expected result:**
            getResumeToken must return the _id of the previous document returned if one exists.
            getResumeToken must return startAfter from the initial aggregate if the option was specified.
            getResumeToken must return resumeAfter from the initial aggregate if the option was specified.
            If neither the startAfter nor resumeAfter options were specified, the getResumeToken result must be empty.

DmitryLukyanov commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs:
```diff
@@ -76,6 +113,16 @@ public void Dispose()
+        public BsonDocument GetResumeToken()
+        {
+            return 
+                _postBatchResumeToken ?? 
```
these lines together with `ChangeStreamOperation.GetInitialPostBatchResumeTokenIfRequired` satisfies to the unit tests result which are proposed for this ticket.

rstam commented on src/MongoDB.Driver.Core/IAsyncCursor.cs:
```diff
@@ -52,6 +53,12 @@ public interface IAsyncCursor<out TDocument> : IDisposable
         Task<bool> MoveNextAsync(CancellationToken cancellationToken = default(CancellationToken));
     }
 
+    internal interface IBatchInfo
```
Can we rename this ICursorBatchInfo?

Just plain "batch info" is too open ended.

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/IAsyncCursor.cs:
```diff
@@ -52,6 +53,12 @@ public interface IAsyncCursor<out TDocument> : IDisposable
+    internal interface IBatchInfo
+    {
+        BsonDocument PostBatchResumeToken { get; }
+        bool IsEmpty { get; }
```
Rename to WasFirstBatchEmpty?

And see comment elsewhere on why this property really only applies to the first batch.

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/AggregateOperation.cs:
```diff
@@ -468,6 +471,11 @@ public override AggregateResult Deserialize(BsonDeserializationContext context,
 
+                        case "postBatchResumeToken":
+                            var bsonSerializer = new BsonDocumentSerializer();
+                            result.PostBatchResumeToken = bsonSerializer.Deserialize(context);
```
You can use the pre-recreated BsonDocumentSerializer:

    result.PostBatchResumeToken = BsonDocumentSerializer.Instance.Deserialize(context);

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/AsyncCursor.cs:
```diff
@@ -87,13 +88,40 @@ public class AsyncCursor<TDocument> : IAsyncCursor<TDocument>
+        {
+        }
+
+        internal AsyncCursor(
```
The other constructor is public, so this one might as well be public also.

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/AsyncCursor.cs:
```diff
@@ -87,13 +88,40 @@ public class AsyncCursor<TDocument> : IAsyncCursor<TDocument>
+            int? limit,
+            IBsonSerializer<TDocument> serializer,
+            MessageEncoderSettings messageEncoderSettings,
+            TimeSpan? maxTime = null)
```
No need for maxTime to default to null.

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs:
```diff
@@ -25,20 +25,41 @@
+    /// Represents an asynchronous cursor for change stream.
+    /// </summary>
+    /// <typeparam name="TDocument"></typeparam>
+    public interface IChangeStreamCursor<out TDocument> : IAsyncCursor<TDocument>
```
It might make more sense to put this interface declaration in IAsyncCursor.cs.

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs:
```diff
@@ -25,20 +25,41 @@
+        private BsonDocument _postBatchResumeToken;
+        private readonly BsonDocument _resumeAfter;
+        private readonly BsonDocument _startAfter;
+        private readonly BsonTimestamp _startAtOperationTime;
```
Rename these three variables to

    _initialResumeAfter
    _initialStartAfter
    _initialStartAtOperationTime

to be very clear that these are the initial values supplied by the application and not necessarily the values that will be used when resuming.

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs:
```diff
@@ -89,9 +136,7 @@ public bool MoveNext(CancellationToken cancellationToken = default(CancellationT
-                    var newCursor = _changeStreamOperation.Resume(_binding, cancellationToken);
-                    _cursor.Dispose();
-                    _cursor = newCursor;
+                    _cursor = ResumeCursor(_cursor, cancellationToken);
```
Do this instead:

    var newCursor = Resume(cancellationToken);
    _cursor.Dispose();
    _cursor = newCursor;

It is better for the call to Dispose to be in the same method where we replace the old value of _cursor with a new one.

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs:
```diff
@@ -112,9 +157,7 @@ public async Task<bool> MoveNextAsync(CancellationToken cancellationToken = defa
-                    var newCursor = await _changeStreamOperation.ResumeAsync(_binding, cancellationToken).ConfigureAwait(false);
-                    _cursor.Dispose();
-                    _cursor = newCursor;
+                    _cursor = await ResumeCursorAsync(_cursor, cancellationToken).ConfigureAwait(false);
```
Same as above.

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs:
```diff
@@ -154,12 +199,54 @@ private IEnumerable<TDocument> DeserializeDocuments(IEnumerable<RawBsonDocument>
             return documents;
         }
 
+        private ResumeStartValue GetEffectiveResumeStartValues()
```
Rename to GetResumeValues.

"Effective" made sense when this method was ChangeStreamOperation, but here in this class the shorter name is better.

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs:
```diff
@@ -154,12 +199,54 @@ private IEnumerable<TDocument> DeserializeDocuments(IEnumerable<RawBsonDocument>
+            return new ResumeStartValue { ResumeAfter = _resumeAfter, StartAfter = _startAfter, StartAtOperationTime = _startAtOperationTime };
+        }
+
+        private BsonDocument GetPostBatchResumeTokenOrThrow()
```
I don't think this method is worth it. Just cast _cursor directly to ICursorBatchInfo when needed.

I don't see any reason we need a custom error message for an InvalidCastException that will never be thrown.

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs:
```diff
@@ -139,6 +182,8 @@ private IEnumerable<TDocument> DeserializeDocuments(IEnumerable<RawBsonDocument>
             var documents = new List<TDocument>();
             RawBsonDocument lastRawDocument = null;
 
+            _postBatchResumeToken = GetPostBatchResumeTokenOrThrow();
```
No need for a helper method, just cast directly:

    _postBatchResumeToken = ((ICursorBatchInfo)_cursor).PostBatchResumeToken;

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs:
```diff
@@ -181,5 +268,29 @@ private void ProcessBatch(bool hasMore)
+            _changeStreamOperation = _changeStreamOperation.WithResumeOptions(_documentResumeToken, _postBatchResumeToken, _initialOperationTime);
+            var newCursor = _changeStreamOperation.Resume(_binding, cancellationToken);
+            cursor.Dispose();
+            return newCursor;
```
I don't think we need to create a new operation for Resume. It takes a lot of code to clone the operation and we might make mistakes now or in the future when cloning. We can just reconfigure the resume values of the existing operation.

I would rename and simplify this method to:

    private IAsyncCursor<RawBsonDocument> Resume(CancellationToken cancellationToken)
    {
        ReconfigureOperationResumeValues();
        return _changeStreamOperation.Resume(_binding, cancellationToken);
    }

Where ReconfigureOperationResumeValues is:

    private void ReconfigureOperationResumeValues()
    {
        var resumeValues = GetResumeValues();
        _changeStreamOperation.ResumeAfter = resumeValues.ResumeAfter;
        _changeStreamOperation.StartAfter = resumeValues.StartAfter;
        _changeStreamOperation.StartAtOperationTime = resumeValues.StartAtOperationTime;
    }

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs:
```diff
@@ -181,5 +268,29 @@ private void ProcessBatch(bool hasMore)
+            _changeStreamOperation = _changeStreamOperation.WithResumeOptions(effectiveOptions.StartAfter, effectiveOptions.ResumeAfter, effectiveOptions.StartAtOperationTime);
+            var newCursor = await _changeStreamOperation.ResumeAsync(_binding, cancellationToken).ConfigureAwait(false);
+            cursor.Dispose();
+            return newCursor;
```
Same as above.

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs:
```diff
@@ -181,5 +268,29 @@ private void ProcessBatch(bool hasMore)
+            return newCursor;
+        }
+
+        internal struct ResumeStartValue
```
Rename to ResumeValues (plural).

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs:
```diff
@@ -74,6 +66,17 @@ public interface IChangeStreamOperation<TResult> : IReadOperation<IAsyncCursor<T
+        /// <returns>
+        /// A new instance of IChangeStreamOperation.
+        /// </returns>
+        IChangeStreamOperation<TResult> WithResumeOptions(BsonDocument startAfter, BsonDocument resumeAfter, BsonTimestamp startAtOperationTime);
```
Remove this WithResumeOptions method and replace it instead with the following property that can be used along with ResumeAfter and StartAtOperationTime to reconfigure the resume values of the existing operation:

    /// <summary>
    /// Gets or sets the start after value.
    /// </summary>
    /// <value>
    /// The start after value.
    /// </value>
    BsonDocument StartAfter { get; set; }

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs:
```diff
@@ -265,7 +259,12 @@ public BsonDocument ResumeAfter
+        /// </summary>
+        /// <value>
+        /// The start after value.
+        /// </value>
```
You can use:

    /// <inheritdoc />

DmitryLukyanov replied:
Currently, yes. Done

rstam replied:
This doesn't seem to be done.

DmitryLukyanov replied:
Sorry, It was accidentally rollbacked somehow. Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs:
```diff
@@ -311,33 +324,77 @@ public async Task<IAsyncCursor<TResult>> ExecuteAsync(IReadBinding binding, Canc
+        }
+
+        /// <inheritdoc />
+        public IChangeStreamOperation<TResult> WithResumeOptions(BsonDocument startAfter, BsonDocument resumeAfter, BsonTimestamp startAtOperationTime)
```
Let's get rid of this method and instead of cloning the operation for Resume let's just reconfigure the three Resume values and re-use the existing operation.

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs:
```diff
@@ -290,19 +289,33 @@ public IAsyncCursor<TResult> Execute(IReadBinding binding, CancellationToken can
-                cursor = ExecuteAggregateOperation(channelBinding, resuming: false, cancellationToken);
-                SaveInitialOperationTimeIfRequired(channel, channelBinding, cursor);
+                cursor = ExecuteAggregateOperation(channelBinding, cancellationToken);
+                batchInfo = GetBatchInfoOrThrow(cursor);
```
No need for a helper method, just use:

    cursorBatchInfo = (ICursorBatchInfo)cursor;

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs:
```diff
@@ -311,33 +324,77 @@ public async Task<IAsyncCursor<TResult>> ExecuteAsync(IReadBinding binding, Canc
-                cursor = await ExecuteAggregateOperationAsync(channelBinding, resuming: false, cancellationToken).ConfigureAwait(false);
-                SaveInitialOperationTimeIfRequired(channel, channelBinding, cursor);
+                cursor = await ExecuteAggregateOperationAsync(channelBinding, cancellationToken).ConfigureAwait(false);
+                batchInfo = GetBatchInfoOrThrow(cursor);
```
Same as above.

DmitryLukyanov replied:
Done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamOperation.cs:
```diff
@@ -382,59 +437,51 @@ private List<BsonDocument> CreateCombinedPipeline(BsonDocument changeStreamStage
         }
 
-        private ResumeStartValue GetEffectiveResumeStartValues(bool resuming)
+        private IBatchInfo GetBatchInfoOrThrow(IAsyncCursor<RawBsonDocument> cursor)
```
No need for this helper method. Just cast directly to ICursorBatchInfo when needed.

DmitryLukyanov replied:
Done

rstam reviewed (changes requested):
This looks really good!

Don't be alarmed by the number of comments, they are mostly just requests to rename things slightly.

rstam commented on src/MongoDB.Driver.Core/Core/Operations/AsyncCursor.cs:
```diff
@@ -59,10 +59,39 @@ public class AsyncCursor<TDocument> : IAsyncCursor<TDocument>
+        private readonly bool _wasFirstBatchEmpty;
 
         // constructors
+        /// <inheritdoc/>
```
I don't think you can use <inheritdoc/> on a constructor.

DmitryLukyanov replied:
done

rstam commented on src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs:
```diff
@@ -30,15 +30,21 @@ namespace MongoDB.Driver.Core.Operations
         // private fields
         private readonly IReadBinding _binding;
-        private readonly IChangeStreamOperation<TDocument> _changeStreamOperation;
+        private IChangeStreamOperation<TDocument> _changeStreamOperation;
```
Can be readonly.

DmitryLukyanov replied:
done

rstam commented on tests/MongoDB.Driver.Core.Tests/Core/Operations/ChangeStreamCursorTests.cs:
```diff
@@ -20,51 +20,103 @@
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Driver.Core.TestHelpers.XunitExtensions;
 using Xunit;
+using Reflector = MongoDB.Bson.TestHelpers.Reflector;
```
Why "Reflector ="? Normally we just use:

    using MongoDB.Bson.TestHelpers;

DmitryLukyanov replied:
Done

rstam reviewed (changes requested):
Only a few minor comments.

Also, there are a few minor issues with doc comments that Visual Studio does not catch. To find them and to later verify that they are fixed run:

    .\build.ps1 -target apidocs

Then unzip a folder named something like:

    ApiDocs-2.9.0-csharp2514-v2-1-html.zip

in the folder:

    <mongo-csharp-driver-solution-folder>\artifacts\docs

And search for "error" and "warning" in the file

    LastBuild.log

DmitryLukyanov commented:
It's strange that this way `.\build.ps1 -target apidocs` to build the app is not the same with how the vs handles docs attributes.
For example processing of `WasFirstBatchEmpty` and `PostBatchResumeToken` don't look expected to me. When I've added summary comments to the base interface, the docs warnings 
>SHFB: Warning GID0004: No comments found for member &#39;P:MongoDB.Driver.Core.Operations.AsyncCursor`1.PostBatchResumeToken&#39;
 
still appeared for this code in `AsyncCursor`:

        /// <inheritdoc/>
        public BsonDocument PostBatchResumeToken
        {
            get { return _postBatchResumeToken; }
        } 

Only when I also added the full field description in `AsyncCursor`, the warning disappeared.

rstam commented:
It is annoying that Sandcastle is stricter about doc comments than Visual Studio and sometimes results in errors or warnings that don't happen in Visual Studio.

The issue with PostBatchResumeToken in AsyncCursor.cs might be that \<inheritdoc/> does not work for a public class to inherit from an internal interface.

In any event your workaround is fine.

rstam reviewed (approved):
Looks really good Dmitry.

jyemin reviewed (approved):
This looks nice.  LGTM!

@rstam how do you want to capture the need for a release note regarding the (minor) backwards breaking change to the return type of the Watch methods?

rstam commented:
@jyemin suggested alternatives:

1. write a note in the placeholder release notes for 2.9.0
2. create a JIRA ticket
